libraries/shared/fetcher/geth_rpc_storage_fetcher.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2019 Vulcanize
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fetcher
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/statediff"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/storage/utils"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/streamer"
+)
+
+type GethRpcStorageFetcher struct{
+	statediffPayloadChan chan statediff.Payload
+	streamer streamer.Streamer
+}
+
+func NewGethRpcStorageFetcher(streamer streamer.Streamer, statediffPayloadChan chan statediff.Payload) GethRpcStorageFetcher{
+	return GethRpcStorageFetcher{
+		statediffPayloadChan: statediffPayloadChan,
+		streamer:  streamer,
+	}
+}
+
+func (fetcher *GethRpcStorageFetcher) FetchStorageDiffs(out chan<- utils.StorageDiffRow, errs chan<- error) {
+	ethStatediffPayloadChan := fetcher.statediffPayloadChan
+	_, err := fetcher.streamer.Stream(ethStatediffPayloadChan)
+	if err != nil {
+		errs <- err
+	}
+
+	for diff := range ethStatediffPayloadChan {
```

Comment:
Should this be a select statement within a repeating for loop (possibly within a goroutine? similar to https://github.com/vulcanize/vulcanizedb/blob/ipfs_concurrency/pkg/ipfs/service.go#L147) As is, once it iterates through everything that is in the channel it will fall through and exit this method, it won't wait for and process new things that are added to the channel.

Reply:
Nice catch! I've added a for loop here instead of a range. I don't _think_ it needs a goroutine, since `FetchStorageDiffs` is being called in a goroutine, but I could be wrong. 🤔

Reply:
Yeah if that's the case there should be no need for an internal goroutine! I may be missing something, but it might be worth enforcing buffering on the `statediffPayloadChan`. I think that channel is only ever used internal to the `FetchStorageDiffs` method so instead of passing in the chan as an argument to `NewGethRpcStorageFetcher` it could be initialized with a size as part of the `NewGethRpcStorageFetcher` function.

Reply:
Yeah, I keep going back an forth on that. I think a big part of the reason I am passing it into `NewStorageWatcher` is so that it's easier to test. But totally be happy to reconsider that - testing channels is hard, I'm having trouble seeing a different way to do that at the moment.

Reply:
That's an excellent point, I wasn't thinking about tests (kinda my crux).